sklearn/metrics/classification.py

```diff
@@ -391,72 +391,269 @@ def jaccard_similarity_score(y_true, y_pred, normalize=True,
-    to the ``accuracy_score``. It differs in the multilabel classification
-    problem.
-
     References
     ----------
     .. [1] `Wikipedia entry for the Jaccard index
            <https://en.wikipedia.org/wiki/Jaccard_index>`_
 
-
     Examples
     --------
     >>> import numpy as np
     >>> from sklearn.metrics import jaccard_similarity_score
-    >>> y_pred = [0, 2, 1, 3]
-    >>> y_true = [0, 1, 2, 3]
-    >>> jaccard_similarity_score(y_true, y_pred)
-    0.5
-    >>> jaccard_similarity_score(y_true, y_pred, normalize=False)
-    2
 
     In the multilabel case with binary label indicators:
 
-    >>> jaccard_similarity_score(np.array([[0, 1], [1, 1]]),\
-        np.ones((2, 2)))
+    >>> jaccard_similarity_score(np.array([[0, 1], [1, 1]]),
+    ... np.ones((2, 2)))
     0.75
+
+    In the multiclass case:
+
+    >>> y_pred = [0, 2, 1, 3]
+    >>> y_true = [0, 1, 2, 3]
+    >>> jaccard_similarity_score(y_true, y_pred, average='macro')
+    0.5
+    >>> jaccard_similarity_score(y_true, y_pred, average='micro')
+    ... # doctest: +ELLIPSIS
+    0.333...
+
+    >>> y_pred = ['ant', 'ant', 'cat', 'cat', 'ant', 'cat']
+    >>> y_true = ['cat', 'ant', 'cat', 'cat', 'ant', 'bird']
+    >>> jaccard_similarity_score(y_true, y_pred, average='weighted')
+    ... # doctest: +ELLIPSIS
+    0.4722...
     """
 
+    average_options = (None, 'micro', 'macro', 'weighted', 'samples',
+                       'none-samples')
+    if average not in average_options and average != 'binary':
+        raise ValueError("average has to be one of " + str(average_options))
+
     # Compute accuracy for each possible representation
     y_type, y_true, y_pred = _check_targets(y_true, y_pred)
     check_consistent_length(y_true, y_pred, sample_weight)
+    present_labels = unique_labels(y_true, y_pred)
+
+    if average == 'binary':
+        if y_type == 'binary':
+            if pos_label not in present_labels:
+                if len(present_labels) < 2:
+                    return 0.
+                else:
+                    raise ValueError("pos_label=%r is not a valid label: "
+                                    "%r" % (pos_label, present_labels))
+            labels = [pos_label]
+        else:
+            raise ValueError("Target is %s but average='binary'. Please "
+                            "choose another average setting." % y_type)
+    elif pos_label not in (None, 1):
+        warnings.warn("Note that pos_label (set to %r) is ignored when "
+                      "average != 'binary' (got %r). You may use "
+                      "labels=[pos_label] to specify a single positive class."
+                      % (pos_label, average), UserWarning)
+
+    if labels is None:
+        labels = present_labels
+        n_labels = None
+    else:
+        n_labels = len(labels)
+        labels = np.hstack([labels, np.setdiff1d(present_labels, labels,
+                                                 assume_unique=True)])
+
```

Comment:
It is very tempting to just handle the multiclass case by applying LabelBinarizer and cutting out lots of code

Reply:
As I can understand, this is just sort of extending the labels to include all *possible* labels classes. So I can't make sense of your comment here, as to Binarizing the multiclass case. I don't think it is doing anything related to binarizing the multiclass input.

As for your comment, from the moment your first mentioned about *invariance* between multiclass and multilabel inputs (for jaccard), I was wondering why isn't it advisable to just binarize the input and passing it again to the multilable input, even all average's supported by multilclass are actually supported by multilabel as well?

Reply:
I wasn't referring to the code that comes before this line, but the code that comes after it.

There's no reason not to just implement it that way, except perhaps for computational performance. Which I wouldn't worry about for now.